## Import dependencies and read csv

In [1]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced
from sklearn.ensemble import RandomForestClassifier

In [ ]:
#credentials for database

#from sqlalchemy import create_engine
#from config_db import db_password
# Read ml_input table into dataframe
db_string = f”postgresql://postgres:{db_password}@127.0.0.1:5432/stackoverflow”
engine = create_engine(db_string)
# table named ‘ml_input’ will be returned as a dataframe.
ml_input_df = pd.read_sql_table(‘ml_input’, engine)
ml_input_df

In [14]:
#csv for testing our ML models

file_path='../Resources/ML_Input_Jan2021.csv'
df=pd.read_csv(file_path)

In [15]:
df.head()

q_id  accepted_answer_id  q_score         q_score_tier  q_view_count  \
0  65526420            65526457        2  Positive Score (>0)            62   
1  65526423            65526533        2  Positive Score (>0)            48   
2  65526490            65526541        2  Positive Score (>0)            35   
3  65526419            65526554        3  Positive Score (>0)           351   
4  65526523            65526577        2  Positive Score (>0)           117   

  q_view_count_bin  q_title_char_count q_title_char_count_bin  \
0         50-16000                  72        Medium (50-100)   
1            40-50                  48         Short (0 - 50)   
2            30-40                  81        Medium (50-100)   
3         50-16000                  76        Medium (50-100)   
4         50-16000                  82        Medium (50-100)   

   q_title_word_count q_title_word_count_bin  q_body_word_count  \
0                  13         Medium (10-20)                116   
1                   8         Short (0 - 10)                 58   
2                  13         Medium (10-20)                117   
3                   9         Short (0 - 10)                 50   
4                  14         Medium (10-20)                305   

  q_body_len_bin  q_tags_count   q_day  q_hour q_hour_min  \
0        100-250             3  Friday       0      00:05   
1         50-100             2  Friday       0      00:06   
2        100-250             2  Friday       0      00:20   
3            <50             4  Friday       0      00:05   
4        250-500             3  Friday       0      00:30   

   accepted_answer_duration  
0                  0.122066  
1                  0.475172  
2                  0.287423  
3                  0.575997  
4                  0.253412

## Data Preprocessing

- Dropped null values and dropped uncessary columns
- Binned data in accepted_answer_duration

In [16]:
#drop identification columns also q_hour_min and q_body_len_bin columns as they are redundant to other columns

df=df.drop(['q_id','accepted_answer_id','q_hour_min','q_score','q_view_count','q_score_tier','q_view_count_bin'], axis=1)
df.head()

q_title_char_count q_title_char_count_bin  q_title_word_count  \
0                  72        Medium (50-100)                  13   
1                  48         Short (0 - 50)                   8   
2                  81        Medium (50-100)                  13   
3                  76        Medium (50-100)                   9   
4                  82        Medium (50-100)                  14   

  q_title_word_count_bin  q_body_word_count q_body_len_bin  q_tags_count  \
0         Medium (10-20)                116        100-250             3   
1         Short (0 - 10)                 58         50-100             2   
2         Medium (10-20)                117        100-250             2   
3         Short (0 - 10)                 50            <50             4   
4         Medium (10-20)                305        250-500             3   

    q_day  q_hour  accepted_answer_duration  
0  Friday       0                  0.122066  
1  Friday       0                  0.475172  
2  Friday       0                  0.287423  
3  Friday       0                  0.575997  
4  Friday       0                  0.253412

In [17]:
df=df.dropna()

In [65]:
#bin accepted_answer_duration

answer_bins = [0, 1, 6000]
answer_bins_group_names = ["<1H", ">1H"]

# Categorize score based on the bins.
df['accepted_answer_duration_bin'] = pd.cut(df['accepted_answer_duration'], answer_bins, labels=answer_bins_group_names)

In [66]:
df.head()

q_title_char_count q_title_char_count_bin  q_title_word_count  \
0                  72        Medium (50-100)                  13   
1                  48         Short (0 - 50)                   8   
2                  81        Medium (50-100)                  13   
3                  76        Medium (50-100)                   9   
4                  82        Medium (50-100)                  14   

  q_title_word_count_bin  q_body_word_count q_body_len_bin  q_tags_count  \
0         Medium (10-20)                116        100-250             3   
1         Short (0 - 10)                 58         50-100             2   
2         Medium (10-20)                117        100-250             2   
3         Short (0 - 10)                 50            <50             4   
4         Medium (10-20)                305        250-500             3   

    q_day  q_hour  accepted_answer_duration accepted_answer_duration_bin  
0  Friday       0                  0.122066                          <1H  
1  Friday       0                  0.475172                          <1H  
2  Friday       0                  0.287423                          <1H  
3  Friday       0                  0.575997                          <1H  
4  Friday       0                  0.253412                          <1H

In [67]:
df['accepted_answer_duration_bin'].value_counts()

<1H    233416
>1H    214088
Name: accepted_answer_duration_bin, dtype: int64

In [68]:
df.dtypes

q_title_char_count                 int64
q_title_char_count_bin            object
q_title_word_count                 int64
q_title_word_count_bin            object
q_body_word_count                  int64
q_body_len_bin                    object
q_tags_count                       int64
q_day                             object
q_hour                             int64
accepted_answer_duration         float64
accepted_answer_duration_bin    category
dtype: object

## Create features and encode our features using pd.get_dummies

In [69]:
# Create our features
X = df.drop(['accepted_answer_duration','accepted_answer_duration_bin','q_title_char_count_bin','q_title_word_count_bin','q_body_len_bin'], axis=1)
X = pd.get_dummies(X)

# Create our target

y = df["accepted_answer_duration_bin"]

X.head()

q_title_char_count  q_title_word_count  q_body_word_count  q_tags_count  \
0                  72                  13                116             3   
1                  48                   8                 58             2   
2                  81                  13                117             2   
3                  76                   9                 50             4   
4                  82                  14                305             3   

   q_hour  q_day_Friday  q_day_Monday  q_day_Saturday  q_day_Sunday  \
0       0             1             0               0             0   
1       0             1             0               0             0   
2       0             1             0               0             0   
3       0             1             0               0             0   
4       0             1             0               0             0   

   q_day_Thursday  q_day_Tuesday  q_day_Wednesday  
0               0              0                0  
1               0              0                0  
2               0              0                0  
3               0              0                0  
4               0              0                0

In [70]:
# Check the balance of our target values
y.value_counts()

<1H    233416
>1H    214088
Name: accepted_answer_duration_bin, dtype: int64

## Split data to training and testing sets

In [71]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=1,stratify=y)

# Random Forest Classifier
## Fit model

In [72]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier

brfc = BalancedRandomForestClassifier(n_estimators=100,random_state=1)
rf = brfc.fit(X_train,y_train)

## Calculate Accuracy

In [73]:
# Calculated the balanced accuracy score
y_pred=rf.predict(X_test)
ba_balanced_forest=balanced_accuracy_score(y_test,y_pred)
ba_balanced_forest

0.5480634963271791

accuracy for binned data less than 1 hr: 0.549306031766825

accuracy for binned data less than 24 hr: 0.5659595317949718
___

accuracy for data excluding binned features: 0.5640773057020267

accuracy for data excluding continuous features: 0.5484520553134237

___

accuracy for binned data less than 1 hr excluding binned features: 0.5480634963271791

## Display Confusion Matrix


In [74]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)

cm_df=pd.DataFrame(cm,
                  index=["Actual <1D", "Actual >1D"],
                  columns=["Predicted <1D", "Predicted >1D"])
cm_df

Predicted <1D  Predicted >1D
Actual <1D          32632          25722
Actual >1D          24785          28737

## Print additional scores for analysis: precision, recall, and f1

In [75]:
#imbalanced classification report
icr_balanced_forest=classification_report_imbalanced(y_test,y_pred)

In [76]:
#Summary of findings

print(f'For the Balanced Random Forest Classifier algortihm, the balanced accuracy score is {ba_balanced_forest}' 
      f'\n\nand the imbalanced classifcation report is:\n\n{icr_balanced_forest}')

For the Balanced Random Forest Classifier algortihm, the balanced accuracy score is 0.5480634963271791

and the imbalanced classifcation report is:

                   pre       rec       spe        f1       geo       iba       sup

        <1H       0.57      0.56      0.54      0.56      0.55      0.30     58354
        >1H       0.53      0.54      0.56      0.53      0.55      0.30     53522

avg / total       0.55      0.55      0.55      0.55      0.55      0.30    111876



## Feature Importance

In [37]:
# List the features sorted in descending order by feature importance
sorted(zip(rf.feature_importances_, X.columns), reverse=True)

[(0.26581440533938167, 'q_body_word_count'),
 (0.21945696793549924, 'q_title_char_count'),
 (0.18330686673251512, 'q_hour'),
 (0.11253601396877579, 'q_title_word_count'),
 (0.05740546325776691, 'q_tags_count'),
 (0.03566203158864189, 'q_view_count_bin_50-16000'),
 (0.010766041698690111, 'q_day_Thursday'),
 (0.010412239982643564, 'q_day_Tuesday'),
 (0.010380043252825268, 'q_day_Wednesday'),
 (0.010195998693345655, 'q_day_Monday'),
 (0.009075162985430294, 'q_view_count_bin_20-30'),
 (0.009040632510808527, 'q_day_Saturday'),
 (0.008649329911996368, 'q_day_Sunday'),
 (0.008431817599653812, 'q_view_count_bin_30-40'),
 (0.007994526947916492, 'q_day_Friday'),
 (0.007983184886605195, 'q_score_tier_Zero Score (0)'),
 (0.006908080918690472, 'q_score_tier_Positive Score (>0)'),
 (0.006093119539042898, 'q_score_tier_Negative Score (<0)'),
 (0.005282073952144683, 'q_view_count_bin_40-50'),
 (0.004654994767301666, 'q_title_word_count_bin_Medium (10-20)'),
 (0.004429088189627105, 'q_title_word_count_

# Easy Ensemble Classifier
## Fit Model

In [38]:
# Train the EasyEnsembleClassifier
from imblearn.ensemble import EasyEnsembleClassifier

ensemble=EasyEnsembleClassifier(n_estimators=100,random_state=1)

eec = ensemble.fit(X_train, y_train)

## Calculate Accuracy

In [39]:
# Calculated the balanced accuracy score
y_pred=eec.predict(X_test)
ba_easy_ensemble=balanced_accuracy_score(y_test,y_pred)
ba_easy_ensemble

0.6286691301503385

## Display Confusion Matrix

In [50]:
# Display the confusion matrix
cm=confusion_matrix(y_test, y_pred)
 
cm_df=pd.DataFrame(cm,
                  index=["Actual <1D", "Actual >1D"],
                  columns=["Predicted <1D", "Predicted >1D"])

cm_df

Predicted <1D  Predicted >1D
Actual <1D          60001          36609
Actual >1D           5545           9700

## Print additional scores for analysis: precision, recall, and f1

In [42]:
#imbalanced classification report
icr_easy_ensemble=classification_report_imbalanced(y_test,y_pred)

In [43]:
print(f'For the Easy Ensemble AdaBoost Classifier algortihm, the balanced accuracy score is {ba_easy_ensemble}' 
      f'\n\nand the imbalanced classifcation report is:\n\n{icr_easy_ensemble}')

For the Easy Ensemble AdaBoost Classifier algortihm, the balanced accuracy score is 0.6286691301503385

and the imbalanced classifcation report is:

                   pre       rec       spe        f1       geo       iba       sup

        <1D       0.92      0.62      0.64      0.74      0.63      0.39     96610
        >1D       0.21      0.64      0.62      0.32      0.63      0.40     15245

avg / total       0.82      0.62      0.63      0.68      0.63      0.39    111855



# Compare and Contrast Random Forest and Easy Ensemble

In [49]:
#randonm forest
print(f'For the Balanced Random Forest Classifier algortihm, the balanced accuracy score is {ba_balanced_forest}' 
      f'\n\nand the imbalanced classifcation report is:\n\n{icr_balanced_forest}')

#easy ensemble
print(f'For the Easy Ensemble AdaBoost Classifier algortihm, the balanced accuracy score is {ba_easy_ensemble}' 
      f'\n\nand the imbalanced classifcation report is:\n\n{icr_easy_ensemble}')

For the Balanced Random Forest Classifier algortihm, the balanced accuracy score is 0.6093991256022591

and the imbalanced classifcation report is:

                   pre       rec       spe        f1       geo       iba       sup

        <1D       0.91      0.61      0.61      0.73      0.61      0.37     96610
        >1D       0.20      0.61      0.61      0.30      0.61      0.37     15245

avg / total       0.81      0.61      0.61      0.67      0.61      0.37    111855

For the Easy Ensemble AdaBoost Classifier algortihm, the balanced accuracy score is 0.6286691301503385

and the imbalanced classifcation report is:

                   pre       rec       spe        f1       geo       iba       sup

        <1D       0.92      0.62      0.64      0.74      0.63      0.39     96610
        >1D       0.21      0.64      0.62      0.32      0.63      0.40     15245

avg / total       0.82      0.62      0.63      0.68      0.63      0.39    111855

